doom/config.org
#+title: Config

* Information
My personal information.
#+begin_src emacs-lisp
(setq user-full-name "wusitee"
      user-mail-address "kekdkkwk@outlook.com")
#+end_src
* Client
** Workspace
Prevent client from creating new workspace.
#+begin_src emacs-lisp
(after! persp-mode
  (setq persp-emacsclient-init-frame-behaviour-override "main"))
#+end_src
** font
#+begin_src emacs-lisp
(setq use-default-font-for-symbols nil)
(setq doom-font (font-spec :family "Maple Mono Normal NF CN" :size 15))
#+end_src
** Transparency
Make client transparent.
#+begin_src emacs-lisp
(set-frame-parameter nil 'alpha-background 90)
(add-to-list 'default-frame-alist '(alpha-background . 90))
#+end_src
** Line number
#+begin_src emacs-lisp
(setq display-line-numbers-type t)
#+end_src
** Theme
#+begin_src emacs-lisp
(setq doom-theme 'doom-horizon)
#+end_src
* Tools
** Spelling check
Set the spelling check to en-US
#+begin_src emacs-lisp
(after! langtool
  (setq langtool-default-language "en-US"))
#+end_src
** gptel
#+begin_src emacs-lisp :tangle no
;; vveai offers an instant OpenAI compatible API
(after! gptel
  (setq gptel-model 'gpt-4o
        gptel-backend
        (gptel-make-openai "vveai"
          :protocol "https"
          :host "api.vveai.com"
          :endpoint "/v1/chat/completions"
          :stream t                      ;optionally nil as Cerebras is instant AI
          :key
          (let ((found (auth-source-search :host "vveai.com" :user "apikey")))
            (if found
                (funcall (plist-get (car found) :secret))
              (error "No authentication entry found for vveai.com with user apikey")))
                                        ;can be a function that returns the key
          :models '(claude-3-7-sonnet-20250219
                    gpt-4o
                    gpt-4.1
                    o3-mini
                    claude-3-7-sonnet-20250219-thinking
                    ))
        gptel-default-mode 'org-mode
        gptel-track-media t)
  (add-hook 'gptel-post-stream-hook 'gptel-auto-scroll)
  (add-hook 'gptel-post-response-functions 'gptel-end-of-response))
#+end_src

#+begin_src emacs-lisp :tangle no
(after! gptel
  (setq
   gptel-model 'gemini-2.5-pro
   gptel-backend (gptel-make-gemini "Gemini"
                   :key
                   (let ((found (auth-source-search :host "gemini" :user "apikey")))
                     (if found
                         (funcall (plist-get (car found) :secret))
                       (error "No authentication entry found for vveai.com with user apikey")))
                   :models '(gemini-2.5-flash-lite
                             gemini-2.5-flash
                             gemini-2.5-pro)
                   :stream t)
   gptel-default-mode 'org-mode
   gptel-track-media t)
  (add-hook 'gptel-post-stream-hook 'gptel-auto-scroll)
  (add-hook 'gptel-post-response-functions 'gptel-end-of-response))
#+end_src
** claude-code-ide
#+begin_src emacs-lisp
(use-package! claude-code-ide
  :config
  (claude-code-ide-emacs-tools-setup)
  (map! :leader
        :prefix ("c" . "code")
        :desc "Claude Code IDE" "'" #'claude-code-ide-menu))
#+end_src
** claude-code-ide-extras
#+begin_src emacs-lisp
(use-package! claude-code-ide-extras
  :after claude-code-ide
  :config
  (claude-code-ide-extras-setup))
#+end_src
** kimi-code-ide
#+begin_src emacs-lisp
(use-package! kimi-code-ide
  :bind ("C-c C-k" . kimi-code-ide-menu)
  :config
  (kimi-code-ide-tools-setup))
#+end_src

** lsp
*** Performance
Follow the [[https://emacs-lsp.github.io/lsp-mode/page/performance/][Performance page]]:
#+begin_src emacs-lisp
(setq gc-cons-threshold 100000000)
(setq read-process-output-max (* 1024 1024)) ;; 1mb
#+end_src
*** Inlay Hints
#+begin_src emacs-lisp
(after! lsp-mode
  (setq lsp-inlay-hint-enable t))
#+end_src
*** Basedpyright
#+begin_src emacs-lisp
(setq lsp-pyright-langserver-command "basedpyright")
#+end_src

** Direnv
#+begin_src emacs-lisp :tangle yes
(add-hook 'after-init-hook 'envrc-global-mode)
#+end_src
** Snippets
*** snippet dir
Set the snippet dirs.
#+begin_src emacs-lisp :tangle no
(setq yas-snippet-dirs '("~/.config/doom/snippets/"))
#+end_src
*** tab conflict
Resolve the tab conflict.
#+begin_src emacs-lisp :tangle no
(defun yas/org-very-safe-expand ()
  (let ((yas/fallback-behavior 'return-nil)) (yas/expand)))
(add-hook 'org-mode-hook
          (lambda ()
            (make-variable-buffer-local 'yas/trigger-key)
            (setq yas/trigger-key [tab])
            (add-to-list 'org-tab-first-hook 'yas/org-very-safe-expand)
            (define-key yas/keymap [tab] 'yas-next-field)))
#+end_src
** Terminal
Using fish for v-term.
#+begin_src emacs-lisp
(setq shell-file-name (executable-find "bash"))
(setq-default vterm-shell (executable-find "fish"))
(setq vterm-timer-delay 0) ;; make vterm smoother
#+end_src
* Completion
** Vertico
Add Chinese pinyin search support to Vertico.
#+begin_src emacs-lisp :tangle no
;; It's good practice to ensure the function you're advising is loaded.
;; The `orderless` feature provides `orderless-regexp`.
(with-eval-after-load 'orderless
  (defun pyin-orderless-regexp (orig-func component)
    "Advise orderless-regexp to support pinyin search."
    (let ((result (funcall orig-func component)))
      ;; pyim-cregexp-build is available if the pyim package is loaded.
      ;; Assuming pyim is loaded, this will add pinyin support.
      (if (fboundp 'pyim-cregexp-build)
          (pyim-cregexp-build result)
        result)))

  (advice-add 'orderless-regexp :around #'pyin-orderless-regexp))
#+end_src

** Copilot
Use copilot when needed.
#+begin_src emacs-lisp
(use-package! copilot
  ;; :hook (prog-mode . copilot-mode)
  :bind (:map copilot-completion-map
              ("<tab>" . 'copilot-accept-completion)
              ("TAB" . 'copilot-accept-completion)
              ("C-TAB" . 'copilot-accept-completion-by-word)
              ("C-<tab>" . 'copilot-accept-completion-by-word)))
#+end_src
* editor
** evil
#+begin_src emacs-lisp :tangle yes
(setq evil-respect-visual-line-mode t)
#+end_src
** Chinese word segmentation
#+begin_src emacs-lisp :tangle yes
;; (add-to-list 'load-path "/home/wste/code/emacs-chinese-word-segmentation")
;; (setq cns-prog "/home/wste/code/emacs-chinese-word-segmentation/cnws")
;; (setq cns-dict-directory "/home/wste/code/emacs-chinese-word-segmentation/cppjieba/dict")

;; (require 'cns nil t)
;; (when (featurep 'cns)
;;   (add-hook 'find-file-hook 'cns-auto-enable))
#+end_src
* lang
** Org
*** Latex preview
\(Latex\)
#+begin_src emacs-lisp
(use-package! org-latex-preview
  :config
  ;; Increase preview width
  (plist-put! org-latex-preview-appearance-options
             :page-width 0.8
             :scale 2.0)
  ;; Use dvisvgm to generate previews
  ;; You don't need this, it's the default:
  (setq org-latex-preview-process-default 'dvisvgm)

  ;; Turn on auto-mode, it's built into Org and much faster/more featured than
  ;; org-fragtog. (Remember to turn off/uninstall org-fragtog.)
  (add-hook 'org-mode-hook 'org-latex-preview-mode)

  ;; Block C-n, C-p etc from opening up previews when using auto-mode
  ;; (setq org-latex-preview-auto-ignored-commands
  ;;       '(next-line previous-line mwheel-scroll
  ;;         scroll-up-command scroll-down-command))

  ;; Enable consistent equation numbering
  (setq org-latex-preview-numbered t)

  ;; Bonus: Turn on live previews.  This shows you a live preview of a LaTeX
  ;; fragment and updates the preview in real-time as you edit it.
  ;; To preview only environments, set it to '(block edit-special) instead
  (setq org-latex-preview-live t)
  ;; Preivew latex when open a file
  (setq org-startup-with-latex-preview t))

  ;; More immediate live-previews -- the default delay is 1 second
  ;; (setq org-latex-preview-live-debounce 0.25))
(defun org-latex-preview-check-health (&optional inter)
  "Inspect the relevent system state and setup.
INTER signals whether the function has been called interactively."
  (interactive (list t))
  ;; Collect information
  (let* ((diag `(:interactive ,inter)))
    (plist-put diag :org-version org-version)
    ;; modified variables
    (plist-put diag :modified
               (let ((list))
                 (mapatoms
                  (lambda (v)
                    (and (boundp v)
                         (string-match "\\`\\(org-latex-\\|org-persist-\\)" (symbol-name v))
                         (or (and (symbol-value v)
                                  (string-match "\\(-hook\\|-function\\)\\'" (symbol-name v)))
                             (and
                              (get v 'custom-type) (get v 'standard-value)
                              (not (equal (symbol-value v)
                                          (eval (car (get v 'standard-value)) t)))))
                         (push (cons v (symbol-value v)) list))))
                 list))
    ;; Executables
    ;; latex processors
    (dolist (processor org-latex-compilers)
      (when-let ((path (executable-find processor)))
        (let ((version (with-temp-buffer
                         (thread-last
                           (concat processor " --version")
                           (shell-command-to-string)
                           (insert))
                         (goto-char (point-min))
                         (buffer-substring (point) (line-end-position)))))
          (push (list processor version path) (plist-get diag :latex-processors)))))
    ;; Image converters
    (dolist (converter '("dvipng" "dvisvgm" "convert"))
      (when-let ((path (executable-find converter)))
        (let ((version (with-temp-buffer
                         (thread-last
                           (concat converter " --version")
                           (shell-command-to-string)
                           (insert))
                         (goto-char (point-min))
                         (buffer-substring (point) (line-end-position)))))
          (push (list converter version path) (plist-get diag :image-converters)))))
    (when inter
      (with-current-buffer (get-buffer-create "*Org LaTeX Preview Report*")
        (let ((inhibit-read-only t))
          (erase-buffer)

          (insert (propertize "Your LaTeX preview process" 'face 'outline-1))
          (insert "\n\n")

          (let* ((latex-available (cl-member org-latex-compiler
                                             (plist-get diag :latex-processors)
                                             :key #'car :test #'string=))
                 (precompile-available
                  (and latex-available
                       (not (member org-latex-compiler '("lualatex" "xelatex")))))
                 (proc-info (alist-get
                             org-latex-preview-process-default
                             org-latex-preview-process-alist))
                 (image-converter (cadr (plist-get proc-info :programs)))
                 (image-converter
                  (cl-find-if
                   (lambda (c)
                     (string= image-converter c))
                   (plist-get diag :image-converters)
                   :key #'car))
                 (image-output-type (plist-get proc-info :image-output-type)))
            (if org-latex-preview-process-precompiled
                (insert "Precompile with "
                        (propertize (map-elt org-latex-precompile-compiler-map
                                             org-latex-compiler)
                                    'face
                                    (list
                                     (if precompile-available
                                         '(:inherit success :box t)
                                       '(:inherit error :box t))
                                     'org-block))
                        " → "))
            (insert "LaTeX Compile with "
                    (propertize org-latex-compiler 'face
                                (list
                                 (if latex-available
                                     '(:inherit success :box t)
                                   '(:inherit error :box t))
                                 'org-block))
                    " → ")
            (insert "Convert to "
                    (propertize (upcase image-output-type) 'face '(:weight bold))
                    " with "
                    (propertize (car image-converter) 'face
                                (list
                                 (if image-converter
                                     '(:inherit success :box t)
                                   '(:inherit error :box t))
                                 'org-block))
                    "\n\n")
            (insert (propertize org-latex-compiler 'face 'outline-3)
                    "\n"
                    (if latex-available
                        (concat
                          (propertize
                           (mapconcat #'identity (map-nested-elt diag `(:latex-processors ,org-latex-compiler))
                                      "\n")
                           'face 'org-block)
                          "\n"
                          (when (and latex-available (not precompile-available))
                            (propertize
                             (format "\nWarning: Precompilation not available with %S!\n" org-latex-compiler)
                             'face 'warning)))
                      (propertize "Not found in path!\n" 'face 'error))
                    "\n")

            (insert (propertize (cadr (plist-get proc-info :programs)) 'face 'outline-3)
                    "\n"
                    (if image-converter
                        (propertize
                         (concat
                          (mapconcat #'identity (cdr image-converter) "\n")
                          "\n")
                         'face 'org-block)
                      (propertize "Not found in path!\n" 'face 'error))
                    "\n")
            ;; dvisvgm version check
            (when (equal (car-safe image-converter)
                         "dvisvgm")
              (let* ((version-string (cadr image-converter))
                     (dvisvgm-ver (progn
                                    (string-match "\\([0-9.]+\\)" version-string)
                                    (match-string 1 version-string))))

                (when (version< dvisvgm-ver "3.0")
                  (insert (propertize
                           (format "Warning: dvisvgm version %s < 3.0, displaymath will not be centered."
                                   dvisvgm-ver)
                           'face 'warning)
                          "\n\n"))
                (unless (string-match-p " RSVG" system-configuration-features)
                  (insert (propertize
                           "Error: Emacs was not compiled with SVG support,
images cannot be displayed with dvisvgm"
                           'face 'error)))))
            ;; png support check
            (when (member (car-safe image-converter)
                          '("dvipng" "convert"))
              (unless (string-match-p " PNG" system-configuration-features)
                (insert (propertize
                         (format "Error: Emacs was not compiled with PNG support,
images cannot be displayed with %s"
                                 (car-safe image-converter))))))
            (when (not (and latex-available image-converter))
              (insert "path: " (getenv "PATH") "\n\n")))
          ;; Settings
          (insert (propertize "LaTeX preview options" 'face 'outline-2)
                  "\n")

          (pcase-dolist (`(,var . ,msg)
                         `((,org-latex-preview-process-precompiled . "Precompilation           ")
                           (,org-latex-preview-numbered . "Equation renumbering     ")
                           (,org-latex-preview-cache  . "Caching with org-persist ")))
            (insert (propertize "• " 'face 'org-list-dt)
                    msg
                    (if var
                        (propertize "ON" 'face '(success bold org-block))
                      (propertize "OFF" 'face '(error bold org-block)))
                    "\n"))
          (insert "\n"
                  (propertize "LaTeX preview sizing" 'face 'outline-2) "\n"
                  (propertize "•" 'face 'org-list-dt)
                  " Page width  "
                  (propertize
                   (format "%S" (plist-get org-latex-preview-appearance-options :page-width))
                   'face '(org-code org-block))
                  "   (display equation width in LaTeX)\n"
                  (propertize "•" 'face 'org-list-dt)
                  " Scale       "
                  (propertize
                   (format "%.2f" (plist-get org-latex-preview-appearance-options :scale))
                   'face '(org-code org-block))
                  "  (PNG pixel density multiplier)\n"
                  (propertize "•" 'face 'org-list-dt)
                  " Zoom        "
                  (propertize
                   (format "%.2f" (plist-get org-latex-preview-appearance-options :zoom))
                   'face '(org-code org-block))
                  "  (display scaling factor)\n\n")
          (insert (propertize "LaTeX preview preamble" 'face 'outline-2) "\n")
          (let ((major-mode 'org-mode))
            (let ((point-1 (point)))
              (insert org-latex-preview-preamble "\n")
              (org-src-font-lock-fontify-block 'latex point-1 (point))
              (add-face-text-property point-1 (point) '(:inherit org-block :height 0.9)))
            (insert "\n")
            ;; Diagnostic output
            (insert (propertize "Diagnostic info (copied)" 'face 'outline-2)
                    "\n\n")
            (let ((point-1 (point)))
              (pp diag (current-buffer))
              (org-src-font-lock-fontify-block 'emacs-lisp point-1 (point))
              (add-face-text-property point-1 (point) '(:height 0.9))))
          (gui-select-text (prin1-to-string diag))
          (special-mode))
        (setq-local
         revert-buffer-function
         (lambda (&rest _)
           (call-interactively #'org-latex-preview-check-health)
           (message "Refreshed LaTeX preview diagnostic")))
        (let ((message-log-max nil))
          (toggle-truncate-lines 1))
        (goto-char (point-min))
        (display-buffer (current-buffer))))
    diag))
#+end_src
*** Latex export
Force page break after toc
#+begin_src emacs-lisp
(setq org-latex-toc-command "\\clearpage \\tableofcontents \\clearpage")
#+end_src
**** tags
#+begin_src emacs-lisp :tangle yes
(setq org-tag-alist '(("noexport") ("ignore")))
#+end_src
*** Latex snippets
#+begin_src elisp
(add-hook! 'org-mode (yas-activate-extra-mode 'latex-mode))
#+end_src
*** Org Roam UI
#+begin_src emacs-lisp
(use-package! websocket
  :after org-roam)

(use-package! org-roam-ui
  :after org-roam ;; or :after org
  ;;         normally we'd recommend hooking orui after org-roam, but since org-roam does not have
  ;;         a hookable mode anymore, you're advised to pick something yourself
  ;;         if you don't care about startup time, use
  ;;  :hook (after-init . org-roam-ui-mode)
  :config
  (setq org-roam-ui-sync-theme t
        org-roam-ui-follow t
        org-roam-ui-update-on-save t
        org-roam-ui-open-on-start t))
#+end_src
*** Capture Templates
Setup Org Roam capture templates.
#+begin_src emacs-lisp
(defun org-roam-template-select-tags ()
  "Prompt and add tags for Org Roam captures"
  (let ((tags (completing-read-multiple "Tag: " (org-roam-tag-completions))))
    (if tags
        (concat "#+filetags: " (org-make-tag-string tags))
      "")))

(after! org-roam
  (setq org-roam-capture-templates
        '(("a" "Anki Card" plain "* ${title}\n%?"
           :target (file+head "${slug}.org" ":PROPERTIES:\n:ANKI_DECK: study \n:END:\n#+title: ${title}\n%(org-roam-template-select-tags)\n")
           :unnarrowed t
           :empty-lines 1))))
#+end_src 
*** Anki editor
#+begin_src emacs-lisp
(use-package! anki-editor
  :after org
  :hook (org-mode . (lambda ()
                      (if (ignore-errors (anki-editor-api-check))
                          (anki-editor-mode))))
  :config
  (map! :map anki-editor-mode-map
        :leader
        :prefix ("n e" . "Anki Editor")
        :desc "Anki editor mode" "e" #'anki-editor-mode
        :desc "Insert note" "i" #'anki-editor-insert-note
        :desc "Push notes" "p" #'anki-editor-push-notes
        :desc "Cloze note" "c" #'anki-editor-cloze-dwim
        :desc "Delete current note" "d" #'anki-editor-delete-note-at-point
        :desc "Browse current note" "b" #'anki-editor-gui-browse
        :desc "Set current note type" "t" #'anki-editor-set-note-type
        :desc "Push new notes" "P" #'anki-editor-push-new-notes)
  (setq org-export-with-broken-links t))
#+end_src
**** Change all the cards' deck to study
#+begin_src emacs-lisp :tangle no
(dolist (file (directory-files-recursively "~/org/roam/" "\\.org$"))
  (let ((buf (find-file-noselect file))
        (changed nil))
    (with-current-buffer buf
      ;; Check and update file-level property
      (when (org-entry-get nil "anki_deck")
        (org-entry-put nil "anki_deck" "study")
        (setq changed t))
      ;; Check and update properties in headings
      (org-map-entries
       (lambda ()
         (when (org-entry-get (point) "anki_deck")
           (org-entry-put (point) "anki_deck" "study")
           (setq changed t)))
       "anki_deck" 'file)
      ;; Save the buffer and push notes if changes were made
      (when changed
        (save-buffer)
        (anki-editor-push-notes 'file)))
    (kill-buffer buf)))
#+end_src
*** Org Roam Export to Publish

#+begin_src emacs-lisp :tangle yes
(defun org-roam-get-files-with-tag (&optional tags)
  "Return a list of org-roam files that contain tags in TAGS.
If no tags are provided, return all files."
  (let ((query (if (null tags)
                   [:select :distinct file
                    :from nodes]
                 [:select :distinct file
                  :from nodes
                  :join tags :on (= id node-id)
                  :where (in tag $v1)])))
    (mapcar #'car (org-roam-db-query query tags))))

(defun org-roam-export (target &optional tags)
  "Copy org-roam files with tags in TAGS to TARGET and generate a new Org Roam database.
If no tags are provided, copy all files."
  (interactive
   (list
    (read-directory-name "Target: ")
    (apply #'vector (completing-read-multiple "Tags: " (org-roam-tag-completions)))))
  (let ((files (org-roam-get-files-with-tag tags))
        (org-roam-directory target)
        (org-roam-db-location (concat target "/org-roam.db")))
    (dolist (file files)
      (copy-file file org-roam-directory t))
    (org-roam-db-sync)))
#+end_src

*** pretty
Make org file beautiful.
#+begin_src emacs-lisp :tangle no
(use-package! org
  :hook
  (org-mode . org-modern-mode)
  (org-agenda-finalize . org-modern-agenda)
  :config
  (setq org-modern-table nil))
#+end_src

#+begin_src emacs-lisp :tangle yes
(setq org-modern-table nil)
(setq org-appear-inside-latex 't)
(setq org-appear-autolinks 't)
(setq org-appear-autoemphasis 't)
(setq org-appear-autosubmarkers 't)
(setq org-appear-autoentities 't)
(setq org-appear-autokeywords 't)
#+end_src
*** org directory
#+begin_src emacs-lisp
(setq org-directory "~/org/")
#+end_src
*** org agenda files
Add research todo and schedule files to org agenda.
#+begin_src emacs-lisp
(after! org
  (add-to-list 'org-agenda-files "~/research/todo.org")
  (add-to-list 'org-agenda-files "~/research/plan/schedule.org")
  (add-to-list 'org-agenda-files "~/research/papers/papers.org"))
#+end_src
*** org super agenda
Group agenda items into sections for better scannability.
#+begin_src emacs-lisp
(use-package! org-super-agenda
  :after org-agenda)
#+end_src
*** org agenda cleanup
Keep dated TODO entries out of the lower global TODO list.
#+begin_src emacs-lisp :tangle yes
(after! org
  (setq org-agenda-custom-commands
        '(("n" "Agenda and de-duplicated TODOs"
           ((agenda "")
            (alltodo ""
                     ((org-agenda-todo-ignore-with-date t))))))))
#+end_src
*** org habit
#+begin_src emacs-lisp :tangle yes
(add-to-list 'org-modules 'org-habit)
#+end_src
*** Line number
Disable line number
#+begin_src emacs-lisp
(setq-hook! org-mode display-line-numbers nil)
#+end_src
*** Table Alignment
Use Valign to align table.
#+begin_src emacs-lisp
(add-hook! org-mode
           (valign-mode)
           (setq valign-fancy-bar 1))
#+end_src
*** image slicing
Use org sliced images mode to make image displayed better
#+begin_src emacs-lisp :tangle yes

(add-hook! org-mode
           (org-sliced-images-mode)
           (setq org-sliced-images-mode 1))

(defun +org--toggle-inline-sliced-images-in-subtree (&optional beg end refresh)
  "Refresh inline sliced image previews in the current heading/tree."
  (let* ((beg (or beg
                  (if (org-before-first-heading-p)

nushell/history.txt
ls
z
cd
ls
cd Documents\code\luogu\
ls
time ls
ls
nvim
ls
z 
pwd
neofetch
ls
dir
ll
ls
scoop
ls
gcc -v
gdb
ls
z
cd
ls
cd Documents\code\python\
ls
cd pathfinding\
ls
scoop update
nvim
ls
time ls
ls
z luogu
l
ll
exit
dir
ls.exe 
ls
zoxide.exe path
zoxide.exe 
zoxide.exe init
sudo scoop
fd
nu.config-path
$nu.config-path
cd C:\Users\Simon\AppData\Roaming\nushell\config.nu
cd C:\Users\Simon\AppData\Roaming\nushell
ls
nvim config.nu
clear
neofetch
zoxide.exe init
zoxide.exe init nu
zoxide.exe init nushell
z
ls
nvim config.nu
vi
vim
nvim
ls 
timeit ls
ls
z luogu
zi
ls
zoxide.exe init nushell
$nu.env-path
nvim config.nu
ls
nvim env.nu
zi
z
$nu.config-path
ls
cd C:\Users\Simon\AppData\Roaming\nushell\
ls
zoxide.exe init nushell
nvim config.nu
zi
z
nvim .zoxide.nu
scoop list | grep nu
nvim env.nu
nvim .zoxide.nu
cowsay
cowsay fuck
$nu.config-path
cd C:\Users\Simon\AppData\Roaming\nushell\
ls
nvim config.nu
nnu
ls
cd.
cd
ls
$nu.config-path
cd AppData\Roaming\nushell\
ls
nvim env.nu
cd
ls
nvim .zoxide.nu
zi
ls
z luogu
ls
timeit ls
ls
timeit ls
man timeit
z persit
z persist
ls
nvim al
nvim alacritty\
zi
ls
scoop list | grep nu
scoop list | grep zoxide
ls
z luogu
ls
z luogu
ls
timeit ls
ls
neofetch
z nu
ls
nvim env.nu
ls
z
ls
z nu
ls
nvim config
nvim config.nu
nvim env.nu
nvim
nushell
nvim
ls
pwsh
ls
z persist
ls
nvim alacritty\
msys2
ls
z luogu
ls
time ls
timeit ls.exe 
timeit ls
timeit ls | grep *.exe
ls
z persist
ls
z
fc list
fclist
fclip.exe 
nvim alacritty\alacritty.yml
scoop search wrap
scoop-search.exe wrap
ls
fd
grep
ls
busybox.exe 
ash
klist.exe 
ls
z persist
ls
neofetch
nvim
msys2
ls
scoop uninstall msys2
sudo scoop uninstall msys2 --global
ls
mingw
10 / 2
10 / n
find 
ps
ps | where cpu > 5
date
date now
date now | date to-table 
sys 
sys | get host
echo.exe 
nvim
ls
nvim 
lss
sl
ls
alacritty.exe 
scoop reset -h
scoop uninstall nushell
sudo scoop reset nushell --global
zi
z
ls
zi
ls
timit ls
timeit ls
ls
grep
ls | grep .
6
neofetch
ls
nvim 
z luogu
ls
z luogu
ls
timeit ls
time ls
timeit ls
timeit ls.exe
l
ls
?
1
1+3+4/2*8
1 + 3 + 4 / 2 * 8
ls
neofetch
sh
pwsh
ls
neofetch
sys
help sort
{b: 4, a: 3, c:1} | sort -v
help help
help
ls
sys
{b: 4, a: 3, c:1} | sort -v
sort
{b: 12,a: b+12,c: a+13} | sort -v
{b: 12,a: 12,c: 13} | sort -v
{b: 12,a: b+12,c: a+13} | sort -v
sort
ls | sorted
ls | sort
z Videos\
ls
ls | sorted
ls | sort
ls | size sort
ls | sort-by size
nvim
ls
z 
ls
$nu.config-path
cp AppData\Roaming\nushell Documents\wusitee\
help cp


doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!
(setq +latex-viewers '(okular)) ;; set the defualt latex viewer to okular
(setq scroll-conservatively 101) ;; make the scroll better
(setq shell-file-name (executable-find "bash")) ;; set excutable shell to bash instead of fish
(setq-default vterm-shell (executable-find "fish")) ;; let v-term to use fish
(setq doom-font (font-spec :family "CaskaydiaCove Nerd Font" :size 17 ))
(setq vterm-timer-delay 0.01)
;; org mode config
(after! org
  (setq org-fontify-quote-and-verse-blocks nil
        org-fontify-whole-heading-line nil
        org-hide-leading-stars nil))
;; bind fuzzy find to SPC f h
(use-package! affe
  :config
  (setq affe-find-command "fd -t file -H")
  (map! :map doom-leader-file-map
        :desc "Fuzzy Find" "F" #'affe-find
        :desc "Fuzzy Find Home" "h" (lambda () (interactive) (affe-find "~"))))
;; activate SQL source code blocks
(org-babel-do-load-languages
 'org-babel-load-languages
 '((sql . t)))

;; use zoxide to change directory
(map!
 :map doom-leader-file-map
 "z" #'cd)

;; prevent client from creating new workspace
(after! persp-mode
  (setq persp-emacsclient-init-frame-behaviour-override "main"))

;; language tool
(after! langtool
  (setq langtool-default-language "en-GB"))

;; set org-mode latex-preview size
(setq org-format-latex-options (plist-put org-format-latex-options :scale 0.7))
;; keep latex preview on in org-mode
(after! org
  (setq org-startup-with-latex-preview t))

(use-package! websocket
  :after org-roam)

(use-package! org-roam-ui
  :after org-roam ;; or :after org
  ;;         normally we'd recommend hooking orui after org-roam, but since org-roam does not have
  ;;         a hookable mode anymore, you're advised to pick something yourself
  ;;         if you don't care about startup time, use
  ;;  :hook (after-init . org-roam-ui-mode)
  :config
  (setq org-roam-ui-sync-theme t
        org-roam-ui-follow t
        org-roam-ui-update-on-save t
        org-roam-ui-open-on-start t))

;; use org-morden
(use-package! org
  :hook (org-mode . org-modern-mode)
  :hook (org-agenda-finalize . org-modern-agenda)
  )

;; enable fragtop
(use-package! org-fragtog
  :hook
  (org-mode . org-fragtog-mode))


;; let emacs use okular as pdf browser
(setq doc-view-continuous t
      doc-view-resolver 'doc-view-okular)

;; force page break after toc
(setq org-latex-toc-command "\\clearpage \\tableofcontents \\clearpage")

;; tried to fix YASnippet templates not working in Latex fragments in org files
;; (add-hook! org-mode (yas-activate-extra-mode 'latex-mode))


;; for transparency
(set-frame-parameter nil 'alpha-background 90)
(add-to-list 'default-frame-alist '(alpha-background . 90))

(setq yas-snippet-dirs '("~/.config/doom/snippets/"))
;; solve the key conflict in while using yasinppets and org-mode
(defun yas/org-very-safe-expand ()
  (let ((yas/fallback-behavior 'return-nil)) (yas/expand)))
(add-hook 'org-mode-hook
          (lambda ()
            (make-variable-buffer-local 'yas/trigger-key)
            (setq yas/trigger-key [tab])
            (add-to-list 'org-tab-first-hook 'yas/org-very-safe-expand)
            (define-key yas/keymap [tab] 'yas-next-field)))


;; Inlay Hints
(after! lsp-mode
  (setq lsp-inlay-hint-enable t))

(setq user-full-name "wusitee"
      user-mail-address "kekdkkwk@outlook.com")

;; use anki-editor
(use-package! anki-editor
  :after org)

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-symbol-font' -- for symbols
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;; (setq doom-font (font-spec :family "Fira Code" :size 15 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-one)

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")

;; using lsp-booster
(defun lsp-booster--advice-json-parse (old-fn &rest args)
  "Try to parse bytecode instead of json."
  (or
   (when (equal (following-char) ?#)
     (let ((bytecode (read (current-buffer))))
       (when (byte-code-function-p bytecode)
         (funcall bytecode))))
   (apply old-fn args)))
(advice-add (if (progn (require 'json)
                       (fboundp 'json-parse-buffer))
                'json-parse-buffer
              'json-read)
            :around
            #'lsp-booster--advice-json-parse)

(defun lsp-booster--advice-final-command (old-fn cmd &optional test?)
  "Prepend emacs-lsp-booster command to lsp CMD."
  (let ((orig-result (funcall old-fn cmd test?)))
    (if (and (not test?)                             ;; for check lsp-server-present?
             (not (file-remote-p default-directory)) ;; see lsp-resolve-final-command, it would add extra shell wrapper
             lsp-use-plists
             (not (functionp 'json-rpc-connection))  ;; native json-rpc
             (executable-find "emacs-lsp-booster"))
        (progn
          (when-let ((command-from-exec-path (executable-find (car orig-result))))  ;; resolve command from exec-path (in case not found in $PATH)
            (setcar orig-result command-from-exec-path))
          (message "Using emacs-lsp-booster for %s!" orig-result)
          (cons "emacs-lsp-booster" orig-result))
      orig-result)))
(advice-add 'lsp-resolve-final-command :around #'lsp-booster--advice-final-command)

;; accept completion from copilot and fallback to company
(use-package! copilot
  :hook (prog-mode . copilot-mode)
  :bind (:map copilot-completion-map
              ("<tab>" . 'copilot-accept-completion)
              ("TAB" . 'copilot-accept-completion)
              ("C-TAB" . 'copilot-accept-completion-by-word)
              ("C-<tab>" . 'copilot-accept-completion-by-word)))

;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.
